### PR TITLE
MAINT, TST refactor pickle imports and tests 

### PR DESCRIPTION
**Pull request by pierreglaser**
This PR is essentially the first commit of #12011. If you'd rather merge #12011 directly, feel free to close it. 
It is a numpy-wide refactor of pickle importing and `pickle.dump(s)` function testing. Most notably: 
* Some `pickle.dumps` tests that I did not catch in #12090 that are now looped over for every pickle protocol 
* Some refactoring about how pickle is imported internally: since we don't want a big conditional block  such as:
```python
if sys.version_info[0] >= 3:
    if sys.version_info[1] in (6, 7):
        try:
            import pickle5 as pickle
        except ImportError:
            import pickle
    else:
        import pickle
else:
    import cPickle as pickle
```
each time we import pickle in python, we import it *once* this way, in `numpy.core.numeric`, and then each time `pickle` is used in other files, we import it from `numpy.core.numeric`. The only time we import `pickle` another way is in `numpy/core/setup.py`

Now, `numpy` should be clean `pickle`-wise, meaning:
* there is no more `np.ndarray.dump(s)` call in the test suite.
* there is a consistent pickle importing behavior
* each time pickle.dumps is tested, it is tested for every protocol.

### CODE REVIEW

**Review by mattip — APPROVED** [review]
I like this as a separate PR

File: numpy/core/tests/test_multiarray.py [hunk 1]
```diff
@@ -3549,8 +3551,18 @@ def test_test_zero_rank(self):
 
 
 class TestPickling(object):
+    def test_highest_available_pickle_protocol(self):
```
Comment (charris) [inline numpy/core/tests/test_multiarray.py]: This test is a bit strange, why is it needed?

File: numpy/core/tests/test_regression.py [hunk 2]
```diff
@@ -39,12 +39,13 @@ def test_mem_empty(self):
...
         # Ticket #16
         a = np.transpose(np.array([[2, 9], [7, 0], [3, 8]]))
-        f = BytesIO()
-        pickle.dump(a, f)
-        f.seek(0)
-        b = pickle.load(f)
-        f.close()
-        assert_array_equal(a, b)
+        for proto in range(2, pickle.HIGHEST_PROTOCOL + 1):
+            f = BytesIO()
```
Comment (charris) [inline numpy/core/tests/test_regression.py]: I think most of these types of tests should be changed to use `BytesIO` as a context manager, but that is out of scope for this PR.

**Review by charris — APPROVED** [review]

**Comment by charris** [comment]
Thanks @pierreglaser .

File: numpy/core/tests/test_multiarray.py (hunk 1 above)
Comment (pierreglaser) [reply numpy/core/tests/test_multiarray.py]: By the time I started writing this set of PRs, it was a sort of meta-test that made sure all pickle protocol test loops were actually including protocol 5 if possible (e.g if pickle5 was installed), especially for the `ndarray` class.
Now things have a bit changed: we loop over protocols in plenty of other test files, yet this test is only ran here. I agree that this test is less relevant now, and it can actually be removed for the sake of clarity and consistency.
